File: backtracking/51.n-queens.py
#
# @lc app=leetcode id=51 lang=python3
#
# [51] N-Queens
#

# @lc code=start
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        def isValid(row: int, col: int, board: List[str]) -> bool:
            for i in range(row):
                if board[i][col] == "Q":
                    return False

            i = row - 1
            j = col - 1
            while i >= 0 and j >= 0:
                if board[i][j] == "Q":
                    return False
                i -= 1
                j -= 1

            i = row - 1
            j = col + 1
            while i >= 0 and j < n:
                if board[i][j] == "Q":
                    return False
                i -= 1
                j += 1

            return True

        def backtracking(curRow: int, board: List[str]) -> None:
            if curRow == n:
                res.append(board[:])
                return

            for curCol in range(n):
                if isValid(curRow, curCol, board):
                    board[curRow] = board[curRow][:curCol] + "Q" + board[curRow][curCol + 1:]
                    backtracking(curRow + 1, board)
                    board[curRow] = board[curRow][:curCol] + "." + board[curRow][curCol + 1:]


        res = []
        chessBoard = ["." * n for _ in range(n)]
        # print(board)
        backtracking(0, chessBoard)
        return res
# @lc code=end

File: Backtracking/51.n-queens.java
/*
 * @lc app=leetcode id=51 lang=java
 *
 * [51] N-Queens
 */
import java.util.ArrayList;
import java.util.List;

// @lc code=start
class Solution {
    private List<List<String>> res = new ArrayList<>();

    private boolean isValid(int row, int col, char[][] board, int n) {
        int i = 0;
        int j = 0;
        for (; i < n; i++) {
            if (board[i][col] == 'Q') {
                return false;
            }
        }

        for (i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }

        for (i = row - 1, j = col + 1; i >= 0 && j < n; i--, j++) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }
        return true;
    }

    private void backtracking(int row, int n, char[][] board) {
        if (row == n) {

            res.add(toBoard(board));
            return;
        }

        for (int col = 0; col < n; col++) {
            if (isValid(row, col, board, n)) {
                board[row][col] = 'Q';
                backtracking(row + 1, n, board);
                board[row][col] = '.';
            }
        }
    }

    private List<String> toBoard(char[][] board) {
        List<String> newBoard = new ArrayList<>();
        for (char[] row: board) newBoard.add(new String(row));
        return newBoard;
    }

    public List<List<String>> solveNQueens(int n) {
        char[][] board = new char[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                board[i][j] = '.';
            }
        }
        backtracking(0, n, board);
        return res;

    }
}
// @lc code=end
